### PR TITLE
Gt/updates to flow routing

### PR DESCRIPTION
**Pull request by gregtucker**
This PR adds LakeMapperBarnes as a possible lake mapper to use with the FlowAccumulator. It also removes the need for duplicate execution of 3 of the stack-related functions.

### CODE REVIEW

**Comment by kbarnhart** [comment]
👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍

**Comment by SiccarPoint** [comment]
Nice! Just a note to say that if you're evolving topography, it's unlikely this is the fill you will want. Although it's MUCH superior to the old algorithms on general topography, in a landscape with or heading towards few depressions, it is considerably worse performing...

File: landlab/components/flow_accum/flow_accumulator.py [hunk 1]
```diff
@@ -957,15 +959,27 @@ def _add_director(self, flow_director):
...
+                "reroute_flow",
+                "surface",
+                "method",
+                "fill_flat",
+                "fill_surface",
+                "redirect_flow_steepest_descent",
+                "reaccumulate_flow",
+                "ignore_overfill",
+                "track_lakes",
+            ]
```
Comment (mdpiper) [inline landlab/components/flow_accum/flow_accumulator.py:982]: Would [__slots__](https://docs.python.org/3/reference/datamodel.html#slots) be useful here? This is a speculative comment, not a review comment.

**Comment by gregtucker** [comment]
Thanks @SiccarPoint and @mdpiper. To address the performance issue, I added a line that tests for the presence of pits (grid.at_node['flow__sink_flag'] is nonzero for any core node) as a condition for running the lake mapper. I also fixed the second/third issue. I don't really understand `__slots__` but would be interested to learn more.

**Comment by SiccarPoint** [comment]
Good idea!

**Comment by SiccarPoint** [comment]
Great idea

File: landlab/components/flow_accum/flow_accumulator.py [hunk 2]
```diff
@@ -1122,8 +1143,8 @@ def accumulate_flow(self, update_flow_director=True, update_depression_finder=Tr
             # At the moment, no depression finders work with to-many, so it
             # lives here
             if self._depression_finder_provided is not None:
-                if update_depression_finder:
-                    self._depression_finder.map_depressions()
+                if update_depression_finder and self.pits_present():
+                    self._depression_finder.update()
```
Comment (kbarnhart) [inline landlab/components/flow_accum/flow_accumulator.py:1147]: This is the problem line for #1276

**Pull request merged**